Neos.Neos/Classes/Domain/Service/WorkspaceService.php

```diff
@@ -0,0 +1,349 @@
+                $workspaceName,
+                $baseWorkspaceName,
+                \Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceTitle::fromString($title->value),
+                \Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceDescription::fromString($description->value),
+                ContentStreamId::create()
+            )
+        );
+
+        // TODO catch exceptions
+
+        if ($ownerId !== null) {
+            $this->dbal->insert(self::TABLE_NAME_WORKSPACE_ROLE, [
+                'content_repository_id' => $contentRepositoryId->value,
+                'workspace_name' => $workspaceName->value,
+                'subject' => 'user:' . $ownerId->value,
+                'role' => 'OWNER',
+            ]);
+        }
+
+        // TODO catch exceptions
+
+        $this->dbal->insert(self::TABLE_NAME_WORKSPACE_METADATA, [
+            'content_repository_id' => $contentRepositoryId->value,
+            'workspace_name' => $workspaceName->value,
+            'title' => $title->value,
+            'description' => $description->value,
+            'classification' => $classification->name,
+        ]);
+
+        return $workspaceName;
+    }
+
+    public function createPersonalWorkspaceForUserIfMissing(ContentRepositoryId $contentRepositoryId, User $user): void
+    {
+        $existingWorkspaceName = $this->findPrimaryWorkspaceNameForUser($contentRepositoryId, $user->getId());
+        if ($existingWorkspaceName !== null) {
+            $this->requireWorkspace($contentRepositoryId, $existingWorkspaceName);
+            return;
+        }
+        $this->createWorkspace(
+            $contentRepositoryId,
+            WorkspaceTitle::fromString($user->getLabel()),
+            WorkspaceDescription::empty(),
+            WorkspaceName::forLive(),
+            $user->getId(),
+            WorkspaceClassification::PERSONAL,
+        );
+    }
+
+    public function getWorkspacePermissionsForUser(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName, User $user): WorkspacePermissions
+    {
+        $workspaceMetadata = $this->getWorkspaceMetadata($contentRepositoryId, $workspaceName);
+
+        $userRoles = $this->userService->getAllRoles($user);
+        $userIsAdmin = array_key_exists('Neos.Neos:Administrator', $userRoles);
+
+        $userWorkspaceRoles = $this->getUserWorkspaceRoles($contentRepositoryId, $workspaceName, $user->getId(), array_keys($userRoles));
+
+        return match ($workspaceMetadata->classification) {
+            WorkspaceClassification::UNKNOWN => WorkspacePermissions::create(
+                read: $userIsAdmin,
+                publish: false,
+                manage: $userIsAdmin,
+            ),
+            WorkspaceClassification::ROOT => WorkspacePermissions::create(
+                read: true,
+                publish: $this->privilegeManager->isPrivilegeTargetGrantedForRoles($userRoles, 'Neos.Workspace.Ui:Backend.PublishAllToLiveWorkspace'),
```

Comment:
i think the privilge should belong to neos/neos or neos/neos-core still and not be moved via https://github.com/neos/neos-development-collection/pull/5118

Reply:
or as alternative of moving this target back to neos, or creating a new empty target for this which will be interpreted here, we could check for the `LivePublisher` role?

`$this->securityContext->hasRole('Neos.Neos:LivePublisher')`

Reply:
yes, I was wondering the same.. There is some mismatch between workspace <-> user/role assignment and the global Flow roles & privileges..
I'll think about it tomorrow!